### ISSUE TITLE
[Question]: ENOSPC: no space left on device, mkdtemp '/tmp/playwright-artifacts-SsGICD'

### ISSUE THREAD

**Issue by badgifter** [microsoft/playwright-java#526]
### Your question

So. I'm running a few 100 browser instances every 10 mins. My code basically cycles browsers to make sure the ram usage doesn't spike to much on browsers with failed processes... Works great, but ran into a new issue...

What are these? 
![image](https://user-images.githubusercontent.com/50094885/126864279-a76bda7b-abfe-44a5-9f79-5572864642e6.png)
''' playwright-artifacts-Q36WuX  playwright_firefoxdev_profile-GKiW5n  Temp-d6c836bb-cc9d-4b1c-b141-0094bed397b2 '''

My 1tb disk is 100% full from all these /tmp files on ubuntu 20... 

I assumed when I call browser.close() is would delete everything associated with browser, but it doesnt? How can I make sure these files get deleted .. ? Is killing disk space

**Comment by yury-s** [comment]
> I assumed when I call browser.close() is would delete everything associated with browser, but it doesnt? How can I make sure these files get deleted .. ? Is killing disk space

Those folders should all be deleted once the browser is closed. Can you share a repro scenario?

**Comment by yury-s** [comment]
Can you also run the test with `DEBUG=pw:browser` and paste the logs here?

**Comment by badgifter** [comment]
Hello, I solved the issue. 

Forgot to close since I couldn't find the link to this post. Found it now. 

Here is what happened. 

I was running playwright inside workers for node. I think I was running 5 nodes with 10 workers each. Each worker would open a single browser instance, and then run as many contexts as needed (usually 1-4 per task). None of that really matters but maybe someone will do something similar. I was finding that if the context on an open browser never closes, then the memory usage can scale up really quickly. Due to some poor error handling (wasn't time to check the 100 some different errors), some contexts would remain open and so we implemented a timeout. If a process didn't close within X time, we would close with a timeout that just threw an error up to the master process. The master process would then wait for other processes in the worker thread to finish, then it would immediately close the browser instance in the worker thread, and immediately terminate the worker thread right after and replace with a new worker thread. 

So, the issue was that. We didnt "await" for the browser to close before we terminated the worker thread. This cause the browser not to be cleaned up properly. Just added await for browser.close() and done. So, if you have a browser running in a worker, you need browser to fully close and clean up before the thread can be terminated. FYI

**Comment by yury-s** [comment]
> I was finding that if the context on an open browser never closes, then the memory usage can scale up really quickly.

Yes, this is expected. You have to close context in order to dispose all the resources it and pages created in it may hold (e.g. network data).


> We didnt "await" for the browser to close before we terminated the worker thread. This cause the browser not to be cleaned up properly.

In that case you likely had all these browsers running in parallel and trying to close concurrently which likely slowed down the process even more. The contract of the method requires you to await for the close to complete before that the directories may not be deleted yet.
